// @crystallize/cli-next/build/core/components/ImportStatus.js
import { Box, Text } from 'ink';
import React from 'react';
import { colors } from '../../config/colors.js';
const areaMap = {
    media: 'Media',
    shapes: 'Shapes',
    grids: 'Grids',
    items: 'Items',
    languages: 'Languages',
    priceVariants: 'Price variants',
    vatTypes: 'VAT types',
    topicMaps: 'Topic maps',
    customers: 'Customers',
    orders: 'Orders',
};
function progressText(progress) {
    const arr = Array.apply(null, Array(20)).map((_) => '-');
    const filled = Array.apply(null, Array(Math.floor(progress * 20))).map((_) => '=');
    arr.splice(0, filled.length, ...filled);
    return arr.join('');
}
export const ImportStatus = ({ status }) => {
    return (React.createElement(React.Fragment, null, Object.keys(status).map((key) => {
        const name = areaMap[key] || key;
        const { progress, warnings } = status[key];
        return (React.createElement(React.Fragment, { key: key },
            React.createElement(Box, { flexDirection: "column" },
                React.createElement(Text, null,
                    React.createElement(Text, { bold: true, color: colors.warning }, "["),
                    React.createElement(Text, { dimColor: true }, progressText(progress)),
                    React.createElement(Text, { bold: true, color: colors.warning }, "]"),
                    ' ',
                    (progress * 100).toFixed(0).padStart(3),
                    "%",
                    ' ',
                    React.createElement(Text, { bold: true, color: colors.info },
                        ' ',
                        "|"),
                    ' ',
                    name)),
            warnings.map((warn, index) => (React.createElement(Box, { key: index, marginLeft: 1 },
                React.createElement(Text, null,
                    "\u26A0 ",
                    warn.message,
                    " (",
                    warn.code,
                    ")"))))));
    })));
};
