// @crystallize/import-utilities/dist/bootstrap-tenant/bootstrapper/items/getItemIdsForItemRelations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getItemIdsForItemRelation = void 0;
const utils_1 = require("../utils");
async function getItemIdsForItemRelation(context, onUpdate, item, itemRelations) {
    var _a, _b;
    const ids = [];
    if (!itemRelations ||
        !((_a = itemRelations === null || itemRelations === void 0 ? void 0 : itemRelations.items) === null || _a === void 0 ? void 0 : _a.map) ||
        typeof ((_b = itemRelations.items) === null || _b === void 0 ? void 0 : _b.map) !== 'function') {
        return ids;
    }
    await Promise.all(itemRelations.items.map(async (itemRelation) => {
        if (typeof itemRelation === 'object' && item) {
            const { itemId } = await (0, utils_1.getItemId)({
                context,
                externalReference: itemRelation.externalReference,
                cataloguePath: itemRelation.cataloguePath,
                language: context.targetLanguage || context.defaultLanguage,
            });
            if (itemId) {
                ids.push(itemId);
            }
            else {
                onUpdate({
                    error: {
                        code: 'CANNOT_HANDLE_ITEM_RELATION',
                        message: `Could not determine an ID for related item "${itemRelation.externalReference
                            ? itemRelation.externalReference
                            : itemRelation.cataloguePath}`,
                        item,
                    },
                });
            }
        }
    }));
    return ids;
}
exports.getItemIdsForItemRelation = getItemIdsForItemRelation;
