// @crystallize/import-utilities/dist/bootstrap-tenant/bootstrapper/items/subscriptionPlanPrincingJsonToInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.subscriptionPlanPrincingJsonToInput = void 0;
const handleJsonPriceToPriceInput_1 = require("./handleJsonPriceToPriceInput");
function subscriptionPlanPrincingJsonToInput(pricing, meteredVaribles) {
    var _a;
    function handleMeteredVariable(mv) {
        var _a, _b;
        const id = (_a = meteredVaribles.find((m) => m.identifier === mv.identifier)) === null || _a === void 0 ? void 0 : _a.id;
        if (!id) {
            throw new Error('Cannot find id for metered variable ' + mv.identifier);
        }
        return {
            id,
            tierType: mv.tierType,
            tiers: (_b = mv.tiers) === null || _b === void 0 ? void 0 : _b.map((t) => ({
                threshold: t.threshold,
                priceVariants: (0, handleJsonPriceToPriceInput_1.handleJsonPriceToPriceInput)({
                    jsonPrice: t.price,
                }),
            })),
        };
    }
    return {
        priceVariants: (0, handleJsonPriceToPriceInput_1.handleJsonPriceToPriceInput)({
            jsonPrice: pricing.price,
        }),
        meteredVariables: ((_a = pricing === null || pricing === void 0 ? void 0 : pricing.meteredVariables) === null || _a === void 0 ? void 0 : _a.length) > 0
            ? pricing.meteredVariables.map(handleMeteredVariable)
            : undefined,
    };
}
exports.subscriptionPlanPrincingJsonToInput = subscriptionPlanPrincingJsonToInput;
