// @crystallize/cli-next/build/core/utils/console.js
import chalk from 'chalk';
import { colors } from '../../config/colors.js';
class Output {
    constructor(quiet) {
        this.isQuiet = quiet;
    }
    set quiet(quiet) {
        this.isQuiet = quiet;
    }
    log(...args) {
        if (!this.isQuiet) {
            console.log(...args);
        }
    }
    error(...args) {
        console.error(...args);
    }
}
export const output = new Output(false);
export const styles = {
    warning: chalk.hex(colors.warning),
    info: chalk.hex(colors.info),
    error: chalk.hex(colors.error),
    highlightColor: chalk.hex(colors.highlight),
    section: (message) => {
        return `
${chalk.gray('-'.repeat(75))}
${chalk.hex(colors.info)(message)}
${chalk.gray('-'.repeat(75))}
        `;
    },
    success: (message) => {
        return `
${chalk.bgGreen(' '.repeat(75))}
${chalk.bgGreen('  ')}${chalk.black.bgGreen(message)}${chalk.bgGreen(' '.repeat(75 - message.length - 2))}
${chalk.bgGreen(' '.repeat(75))}
        `;
    },
    failure: (message) => {
        return `
${chalk.bgRed(' '.repeat(75))}
${chalk.bgRed('  ')}${chalk.black.bgRed(message)}${chalk.bgRed(' '.repeat(75 - message.length - 2))}
${chalk.bgRed(' '.repeat(75))}
        `;
    },
    title: (message) => {
        return styles.info(` 
==== ${message} ====
${chalk.gray('-'.repeat(10 + message.length))}`);
    },
    highlight: (message) => {
        return chalk.hex(colors.highlight)(message);
    },
};
