// @crystallize/cli-next/build/core/use-cases/createTenant.js
import { createClient } from '@crystallize/js-api-client';
import { jsonToGraphQLQuery } from 'json-to-graphql-query';
async function cleanShapes(apiClient, tenantId, ...shapeIdentifiers) {
    const mutation = {
        shape: shapeIdentifiers.reduce((memo, shapeIdentifier) => {
            const camelCaseIdentifier = shapeIdentifier.replace(/-([a-z])/g, (g) => g[1].toUpperCase());
            return {
                ...memo,
                [camelCaseIdentifier]: {
                    __aliasFor: 'delete',
                    __args: {
                        identifier: shapeIdentifier,
                        tenantId,
                    },
                },
            };
        }, {}),
    };
    const query = jsonToGraphQLQuery({ mutation });
    const result = await apiClient.pimApi(query);
    return result.shape;
}
async function initiateTenantInCrystallize(apiClient, input) {
    const result = await apiClient.pimApi(`mutation CREATE_TENANT ($identifier: String!, $name: String!) {
            tenant {
                create(input: {
                    identifier: $identifier,
                    isActive: true,
                    name: $name,
                    meta: {
                        key: "cli"
                        value: "yes"
                    }
                }) {
                    id
                    identifier
                }
            }
        }`, {
        name: input.identifier,
        identifier: input.identifier,
    });
    return result.tenant.create;
}
export default async (tenant, credentials) => {
    const apiClient = createClient({
        tenantIdentifier: tenant.identifier,
        accessTokenId: credentials.ACCESS_TOKEN_ID,
        accessTokenSecret: credentials.ACCESS_TOKEN_SECRET,
    });
    const { id } = await initiateTenantInCrystallize(apiClient, tenant);
    await cleanShapes(apiClient, id, 'default-product', 'default-folder', 'default-document').catch(console.error);
};
