// @crystallize/cli-next/build/core/use-cases/fetchTips.js
import { createCatalogueFetcher, createClient } from '@crystallize/js-api-client';
export const staticTips = [
    {
        title: 'Want to learn more about Crystallize? Check out',
        url: 'https://crystallize.com/learn',
        type: '',
    },
];
export default async () => {
    const apiClient = createClient({
        tenantIdentifier: 'crystallize_marketing',
    });
    const fetcher = createCatalogueFetcher(apiClient);
    const tree = {
        subtree: {
            __args: {
                first: 10,
            },
            edges: {
                node: {
                    name: true,
                    path: true,
                },
            },
        },
    };
    const query = {
        blogPosts: {
            __aliasFor: 'catalogue',
            __args: {
                path: '/blog',
                language: 'en',
            },
            ...tree,
        },
        comics: {
            __aliasFor: 'catalogue',
            __args: {
                path: '/comics',
                language: 'en',
            },
            ...tree,
        },
    };
    const data = await fetcher(query).catch(console.error);
    return [
        ...staticTips,
        ...data.blogPosts.subtree.edges.map(({ node }) => ({
            title: node.name,
            url: `https://crystallize.com${node.path}`,
            type: 'blogPost',
        })),
        ...data.comics.subtree.edges.map(({ node }) => ({
            title: node.name,
            url: `https://crystallize.com${node.path}`,
            type: 'comic',
        })),
    ];
};
