// @crystallize/cli-next/build/core/journeys/import-tenant-dump/reducer.js
export function Reducer(state, action) {
    switch (action.type) {
        case 'SHOW_NEW_FEEDBACK':
            return {
                ...state,
                feedbackIndex: (state.feedbackIndex + 1) % action.length,
            };
        case 'IMPORT_STARTED':
            return {
                ...state,
                isImporting: true,
                isDone: false,
            };
        case 'IMPORT_DONE':
            return {
                ...state,
                isImporting: false,
                isDone: true,
            };
        case 'SET_CREDENTIALS': {
            return {
                ...state,
                credentials: action.credentials,
            };
        }
        case 'CHANGE_TENANT': {
            if (state.tenant?.identifier === action.item.identifier) {
                return state;
            }
            return {
                ...state,
                tenant: action.item,
            };
        }
        case 'ADD_MESSAGE': {
            return {
                ...state,
                messages: [...state.messages, `${action.message}`],
            };
        }
        default: {
            throw new Error('Unhandled action type');
        }
    }
}
